torch/_refs/__init__.py

```diff
@@ -2938,6 +2939,48 @@ def diagonal(
     return result
 
 
+@register_decomposition(torch.ops.aten.diag_embed)
+def diag_embed(
+    self: TensorLikeType,
+    offset: int = 0,
+    dim1: int = -2,
+    dim2: int = -1,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.diag_embed
+    """
+    num_dims = self.dim() + 1
+    dim1 = utils.canonicalize_dim(idx=dim1, rank=num_dims)
+    dim2 = utils.canonicalize_dim(idx=dim2, rank=num_dims)
+
+    check(
+        dim1 != dim2, lambda: f"diagonal dimensions cannot be identical {dim1}, {dim2}"
+    )
+
+    import builtins
+
+    new_dim_len = builtins.abs(offset) + list(self.size())[-1]
+
+    sizes = list(self.size())
+
+    sizes.pop()
+    sizes.insert(min(dim1, dim2), new_dim_len)
+    sizes.insert(max(dim1, dim2), new_dim_len)
+
+    result = torch.zeros(
+        sizes,
+        dtype=self.dtype,
+        device=self.device,  # TODO: layout=self.layout
+        requires_grad=self.requires_grad,
+    )
+
+    diag = result.diagonal(offset, dim1, dim2)
+    # Note: the inplace copy_ method breaks forward dtype tests
+    copy_to(diag, self)
```

Comment:
Why not using the index trick here and then implement `torch.eye` in terms of this one passing a vector of all ones, as discussed?
With the right strides, it could be a tensor of 1 element broadcasted to be a tensor of the right size.